src/bmp/encoder.rs

```diff
@@ -33,8 +33,15 @@ impl<'a, W: Write + 'a> BMPEncoder<'a, W> {
 
         let (dib_header_size, written_pixel_size, palette_color_count) = get_pixel_info(c)?;
         let row_pad_size = (4 - (width * written_pixel_size) % 4) % 4; // each row must be padded to a multiple of 4 bytes
-
-        let image_size = width * height * written_pixel_size + (height * row_pad_size);
+        let image_size = width.checked_mul(height);
+        if image_size.is_none() {
+            return Err(ImageError::DimensionError);
+        }
+        let image_size = image_size.unwrap().checked_mul(written_pixel_size);
+        if image_size.is_none() {
+            return Err(ImageError::DimensionError);
+        }
+        let image_size = image_size.unwrap() + (height * row_pad_size);
```

Comment:
Instead of bool `is_none` checks and `unwrap`, it is more idiomatic to use `match` or the `?` (called try) operator. The former is for cases where error handling needs explicit control flow, the latter should be sufficient here. Also, it is easy to chain these operators together so there are no intermediate results that might get accidentally used later (e.g. if one deletes the last assignment during a refactoring).

A sketch of the lines here with this style looks like so:
```
let image_size = width.checked_mul(height)
    .ok_or(ImageError::DimensionError)?
    .checked_mul(written_pixel_size)
    .ok_or(ImageError::DimensionError)?;
```


And finally, note both the computation part `+ (height * row_pad_size)` and `width * written_pixel_size` during padding calcuation itself could also overflow. Fixing these as well would make the code even more stable.

Reply:
Thanks, I knew there's got to be a better way to do it, I got confused by checked_mul returning an option instead of a result so I wasn't aware that I could map from none using ok_or right away 👍 .

Updated the other addition as well but didn't check the inner most multiplication as if height * width didn't trigger overflow it's very unlikely that height * [0-3] would.

Thanks 😄